src/txn/txn_timestamp.c

```diff
@@ -914,46 +953,70 @@ __wt_txn_set_timestamp(WT_SESSION_IMPL *session, const char *cfg[])
+        else if (WT_STRING_MATCH("durable_timestamp", ckey.str, ckey.len))
             WT_RET(__wt_txn_parse_timestamp(session, "durable", &durable_ts, &cval));
-            set_ts = true;
-        } else if (WT_STRING_MATCH("prepare_timestamp", ckey.str, ckey.len)) {
+        else if (WT_STRING_MATCH("prepare_timestamp", ckey.str, ckey.len))
             WT_RET(__wt_txn_parse_timestamp(session, "prepare", &prepare_ts, &cval));
-            set_ts = true;
-        } else if (WT_STRING_MATCH("read_timestamp", ckey.str, ckey.len)) {
+        else if (WT_STRING_MATCH("read_timestamp", ckey.str, ckey.len))
             WT_RET(__wt_txn_parse_timestamp(session, "read", &read_ts, &cval));
-            set_ts = true;
-        }
     }
     WT_RET_NOTFOUND_OK(ret);
 
-    /* Look for a commit timestamp. */
-    if (commit_ts != WT_TS_NONE)
-        WT_RET(__wt_txn_set_commit_timestamp(session, commit_ts));
-
-    /*
-     * Look for a durable timestamp. Durable timestamp should be set only after setting the commit
-     * timestamp.
-     */
-    if (durable_ts != WT_TS_NONE)
-        WT_RET(__wt_txn_set_durable_timestamp(session, durable_ts));
-    __wt_txn_publish_durable_timestamp(session);
+    return (__txn_set_timestamp_internal(session, commit_ts, durable_ts, prepare_ts, read_ts));
+}
 
-    /* Look for a read timestamp. */
-    if (read_ts != WT_TS_NONE)
-        WT_RET(__wt_txn_set_read_timestamp(session, read_ts));
+/*
+ * __ts_txn_type_to_str --
+ *     Convert a WT_TS_TXN_TYPE to a printable string.
+ */
+static const char *
+__ts_txn_type_to_str(WT_TS_TXN_TYPE which)
+{
+    switch (which) {
+    case WT_TS_TXN_TYPE_COMMIT:
+        return "commit_timestamp";
+    case WT_TS_TXN_TYPE_DURABLE:
+        return "durable_timestamp";
+    case WT_TS_TXN_TYPE_PREPARE:
+        return "prepare_timestamp";
+    case WT_TS_TXN_TYPE_READ:
+        return "read_timestamp";
+    }
 
-    /* Look for a prepare timestamp. */
-    if (prepare_ts != WT_TS_NONE)
-        WT_RET(__wt_txn_set_prepare_timestamp(session, prepare_ts));
+    return "UNKNOWN";
+}
 
-    /* Timestamps are only logged in debugging mode. */
-    if (set_ts && FLD_ISSET(conn->log_flags, WT_CONN_LOG_DEBUG_MODE) &&
-      FLD_ISSET(conn->log_flags, WT_CONN_LOG_ENABLED) && !F_ISSET(conn, WT_CONN_RECOVERING))
-        WT_RET(__wt_txn_ts_log(session));
+/*
+ * __wt_txn_set_timestamp_uint --
+ *     Directly set the commit timestamp in a transaction, bypassing parsing logic. Prefer this to
+ *     __wt_txn_set_timestamp when string parsing is a performance bottleneck.
+ */
+int
+__wt_txn_set_timestamp_uint(WT_SESSION_IMPL *session, WT_TS_TXN_TYPE which, wt_timestamp_t ts)
+{
+    WT_RET(__wt_txn_context_check(session, true));
+
+    if (ts != WT_TS_NONE) {
+        switch (which) {
+        case WT_TS_TXN_TYPE_COMMIT:
+            WT_RET(__txn_set_timestamp_internal(session, ts, WT_TS_NONE, WT_TS_NONE, WT_TS_NONE));
```

Comment:
This manual matching up of parameters makes me a bit nervous. Did you think of going the other way here, and having `txn_set_timestamp_internal` take the `WT_TS_TXN_TYPE` and only accept one timestamp at a time?

That would involve refactoring `__wt_txn_set_timestamp` to call `txn_set_timestamp_internal` each time a timestamp is parsed out, rather than as a group at the end of the function.

OTOH: it would have the benefit of meaning that both our APIs work in exactly the same way.

Reply:
Yes, I considered it but it's a behaviour change since the durable and commit timestamps will be published regardless. So when updating multiple timestamps in one call, there's now a time window where `txn->pinned_durable_timestamp` is the commit timestamp instead of the durable timestamp.

Reply:
It took me a while to catch up, I think it's more than that - the commit would be published instead of the durable timestamp (once there is a published timestamp it isn't updated, from what I can tell).
I think we need to fix the code here, to give consistency regardless of the API being used.